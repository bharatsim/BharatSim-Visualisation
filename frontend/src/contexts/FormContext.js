import React, { useContext } from 'react';

const FormContext = React.createContext({});

const FormProvider = FormContext.Provider;

const useFormContext = () => useContext(FormContext);

export { FormContext, FormProvider, useFormContext };
