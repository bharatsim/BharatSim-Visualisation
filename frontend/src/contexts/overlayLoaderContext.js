import React from 'react';

const overlayLoaderContext = React.createContext({
  stopLoader: null,
  startLoader: null,
  showError: null,
  isLoading: false,
});

const OverlayLoaderContextProvider = overlayLoaderContext.Provider;

export { overlayLoaderContext, OverlayLoaderContextProvider };
