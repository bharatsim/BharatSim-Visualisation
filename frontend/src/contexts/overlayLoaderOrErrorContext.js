import React from 'react';

const overlayLoaderOrErrorContext = React.createContext({
  stopLoader: null,
  startLoader: null,
  showError: null,
  isLoading: false,
});

const OverlayLoaderOrErrorContextProvider = overlayLoaderOrErrorContext.Provider;

export { overlayLoaderOrErrorContext, OverlayLoaderOrErrorContextProvider };
