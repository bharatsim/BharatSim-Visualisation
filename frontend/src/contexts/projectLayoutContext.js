import React from 'react';

const projectLayoutContext = React.createContext({
  projectMetadata: {
    id: undefined,
    name: undefined,
  },
  selectedDashboardMetadata: {
    name: undefined,
    _id: undefined,
  },
  addDashboard: null,
  deleteDashboard: null,
});

const ProjectLayoutProvider = projectLayoutContext.Provider;

export { projectLayoutContext, ProjectLayoutProvider };
