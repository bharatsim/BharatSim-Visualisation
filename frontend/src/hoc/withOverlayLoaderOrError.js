import React, { useState } from 'react';
import { OverlayLoaderContextProvider } from '../contexts/overlayLoaderContext';
import OverlayLoader from '../uiComponent/OverlayLoader';
import OverlayError from '../uiComponent/OverlayError';

function withOverlayLoaderOrError(WrappedComponent) {
  return (props) => {
    const [isLoading, setLoading] = useState(false);
    const [error, setError] = useState(false);

    function startLoader() {
      setLoading(true);
    }

    function stopLoader() {
      setLoading(false);
    }

    function showError() {
      setError(true);
    }

    function hideError() {
      setError(false);
    }

    return (
      <OverlayLoaderContextProvider value={{ stopLoader, startLoader, showError }}>
        <WrappedComponent {...props} />
        {isLoading && <OverlayLoader isLoading={isLoading} />}
        {error && <OverlayError isError={error} onClose={hideError} />}
      </OverlayLoaderContextProvider>
    );
  };
}

export default withOverlayLoaderOrError;
