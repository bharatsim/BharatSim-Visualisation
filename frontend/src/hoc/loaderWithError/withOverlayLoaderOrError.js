import React, { useState } from 'react';
import { OverlayLoaderOrErrorContextProvider } from '../../contexts/overlayLoaderOrErrorContext';
import OverlayLoader from '../../uiComponent/OverlayLoader';
import OverlayError from '../../uiComponent/OverlayError';
import {useDispatch, useSelector} from "react-redux";
import * as actions from "./actions";

function withOverlayLoaderOrError(WrappedComponent) {
  return (props) => {
    const loaderCount = useSelector((state)=> state.loaderOrError.loaderCount)
    const isError =  useSelector((state)=> state.loaderOrError.isError)
    const errorConfig =  useSelector((state)=> state.loaderOrError.errorConfig)

    const dispatch =  useDispatch()
    function startLoader() {
      dispatch(actions.startLoader())
    }

    function stopLoader() {
      dispatch(actions.stopLoader())
    }

    function showError(errorConfigs) {
      dispatch(actions.showErrors(errorConfigs))
    }

    function hideError() {
      dispatch(actions.hideErrors())
    }
    return (
      <OverlayLoaderOrErrorContextProvider value={{ stopLoader, startLoader, showError }}>
        <WrappedComponent {...props} />
        {loaderCount >= 1 && <OverlayLoader isLoading={loaderCount >= 1} />}
        {isError && (
          <OverlayError
            isError={isError}
            errorConfig={errorConfig}
            hideError={hideError}
          />
        )}
      </OverlayLoaderOrErrorContextProvider>
    );
  };
}

export default withOverlayLoaderOrError;
