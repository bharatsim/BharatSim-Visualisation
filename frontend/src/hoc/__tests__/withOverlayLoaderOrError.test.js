import React, { useContext } from 'react';
import { render, within, fireEvent } from '@testing-library/react';
import Button from '@material-ui/core/Button';
import { waitFor } from '@testing-library/dom';
import withOverlayLoaderOrError from '../withOverlayLoaderOrError';
import { overlayLoaderContext } from '../../contexts/overlayLoaderContext';

function DummyComponent() {
  const { showError, stopLoader, startLoader } = useContext(overlayLoaderContext);
  return (
    <div>
      <Button onClick={startLoader}>startLoader</Button>
      <Button onClick={stopLoader}>stopLoader</Button>
      <Button onClick={showError}>showError</Button>
      Dummy Component
    </div>
  );
}

describe('withOverlayLoaderOrError', () => {
  const Component = withOverlayLoaderOrError(DummyComponent);
  it('should provide overlay loader error context to child component', () => {
    const { container } = render(<Component />);
    expect(container).toMatchSnapshot();
  });
  it('should start loader from the overlay loader error context', () => {
    const { getByText } = render(<Component />);
    const startLoaderButton = getByText('startLoader');

    fireEvent.click(startLoaderButton);
    const loaderComponent = document.querySelector('svg');

    expect(loaderComponent).toBeInTheDocument();
  });

  it('should stop loader from the overlay loader error context', () => {
    const { getByText } = render(<Component />);
    const stopLoaderButton = getByText('stopLoader');
    const startLoaderButton = getByText('startLoader');

    fireEvent.click(startLoaderButton);
    document.querySelector('svg');

    fireEvent.click(stopLoaderButton);
    const loaderComponentAfterStop = document.querySelector('svg');

    // TODO: search by loading message when added
    expect(loaderComponentAfterStop).toBeNull();
  });

  it('should show error from the overlay loader error context', () => {
    const { getByText } = render(<Component />);
    const showErrorButton = getByText('showError');

    fireEvent.click(showErrorButton);
    const { getByText: getByTextFromModal } = within(document.querySelector('.MuiPaper-root'));

    expect(getByTextFromModal('Error message')).toBeInTheDocument();
  });

  it('should hide error from the overlay loader error context', async () => {
    const { getByText } = render(<Component />);
    const showErrorButton = getByText('showError');

    fireEvent.click(showErrorButton);
    const { getByText: getByTextFromModal } = within(document.querySelector('.MuiPaper-root'));

    const closeErrorButton = getByTextFromModal('Close');
    fireEvent.click(closeErrorButton);

    expect(document.querySelector('.MuiDialog-container')).toBeNull();
  });
});
