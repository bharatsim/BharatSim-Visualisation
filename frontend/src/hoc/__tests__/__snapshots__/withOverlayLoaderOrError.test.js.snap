// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withOverlayLoaderOrError should provide overlay loader error context to child component 1`] = `
<div>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        startLoader
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        stopLoader
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        showError
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    Dummy Component
  </div>
</div>
`;
